src/shared/mount-util.c

```diff
@@ -249,9 +249,11 @@ int bind_remount_recursive_with_mountinfo(
                         if (!d)
                                 return -ENOMEM;
 
-                        r = hashmap_ensure_put(&todo, &path_hash_ops_free, d, ULONG_TO_PTR(flags));
-                        if (r == -EEXIST)
-                                continue;
+                        /* If the same path was recorded, but with different mount flags, update it:
+                         * it means a mount point is overmounted, and libmount returns the "bottom" (or
+                         * older one) first, but we want to reapply the flags from the "top" (or newer
+                         * one). See: https://github.com/systemd/systemd/issues/20032 */
+                        r = hashmap_ensure_replace(&todo, &path_hash_ops_free, d, ULONG_TO_PTR(flags));
```

Comment:
`hashmap_replace()` always take both key and value. So, if the path `d` is already stored in the hashmap `todo`, then the previous path is leaked, and moreover, stored `d` will soon freed...

Not tested, I guess this causes use-after-free.

`hashmap_replace()` is hard to use for me, while `hashmap_update()` is simple.